src/mkdocstrings/handlers/__init__.py

```diff
@@ -123,7 +130,13 @@ def __init__(self, directory: str, theme: str) -> None:
             else:
                 raise ThemeNotSupported(theme)
 
-        self.env = Environment(autoescape=True, loader=FileSystemLoader(str(theme_dir)))
+        dirs = [str(theme_dir)]
+
+        # If the custom directory exists, we need to prepend it to the list:
+        if custom_dir is not None and custom_dir.exists():
+            dirs = [(str(custom_dir))] + dirs
+
+        self.env = Environment(autoescape=True, loader=FileSystemLoader(dirs))
```

Comment:
Instead of all these checks, lets simply pass the three directories (custom one, theme one, fallback one) to the env loader.

Something like

```python
paths = []

if custom_templates is not None:
    paths.append(Path(custom_templates) / directory / theme)

themes_dir = Path(__file__).parent.parent / "templates" / directory

paths.append(str(themes_dir / theme))
paths.append(str(themes_dir / self.FALLBACK_THEME))

self.env = Environment(autoescape=True, loader=FileSystemLoader(paths))
```

Then we can catch the `jinja2.TemplateNotFound` exception in the extension itself.